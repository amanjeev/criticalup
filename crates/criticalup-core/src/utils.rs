// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::errors::WriteFileError;
use sha2::{Digest, Sha256};
use std::hash::Hasher;
use std::path::Path;
use tokio::fs::File;
use tokio::io::BufWriter;

pub(crate) async fn open_file_for_write(path: &Path) -> Result<BufWriter<File>, WriteFileError> {
    // Ensure the parent directory is always present
    if let Some(parent) = path.parent() {
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(WriteFileError::CantCreateParentDirectory)?;
    }

    Ok(BufWriter::new(
        File::create(path).await.map_err(WriteFileError::Io)?,
    ))
}

/// A `Hasher` helper type which is a wrapper to a choice of cryptographic hashing algorithm
/// to generate cryptographic hash of our types. This is needed to make sure we
/// 1. do not use [`DefaultHasher`], which may change its algorithm, for hash state
/// 2. bridge the gap between normal [`Hash`] and cryptographic hash (e.g. [`Sha256`])
/// 3. better ergonomics to create a hash of our types like [`ProjectManifestProduct`] using `#[derive(Hash)]`
pub struct Sha256Hasher(Sha256);

impl Sha256Hasher {
    pub(crate) fn new() -> Self {
        Self(Sha256::new())
    }

    /// Provides the final hash value
    pub(crate) fn finalize(self) -> String {
        format!("{:x}", self.0.finalize())
    }
}

impl Hasher for Sha256Hasher {
    /// This method is unreachable and here to appease the compiler, mandatory method.
    fn finish(&self) -> u64 {
        unreachable!()
    }

    /// Update the hasher state, mandatory method.
    fn write(&mut self, bytes: &[u8]) {
        self.0.update(bytes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::hash::Hash;

    #[test]
    fn test_sha256_for_a_struct_works() {
        #[derive(Hash)]
        struct Abc {
            name: String,
            version: String,
            dependencies: Vec<String>,
        }

        let abc = Abc {
            name: "abc".to_string(),
            version: "1.2.3".to_string(),
            dependencies: vec!["dep2".to_string(), "dep1".to_string()],
        };

        let mut hasher = Sha256Hasher::new();
        abc.hash(&mut hasher);
        let final_hash = hasher.finalize();
        assert_eq!(
            "fb9eee112b5cee551f7a5088402e53dbbdaec2a1e1cd3f4663a1b81b1b53015f",
            final_hash
        )
    }
}
