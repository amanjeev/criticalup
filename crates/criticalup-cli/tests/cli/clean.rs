// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::assert_output;
use crate::utils::TestEnvironment;
use serde_json::{json, Value};
use std::fs;
use std::fs::File;
use std::io::{BufReader, Write};

#[test]
fn help_message() {
    let env = TestEnvironment::prepare();
    assert_output!(env.cmd().args(["clean", "--help"]));
}

#[test]
fn clean_deletes_only_unused_installations() {
    let test_env = TestEnvironment::prepare();

    let installation_id_1 = "installation_id_1";
    let installation_id_2 = "installation_id_2";
    let installation_id_3 = "installation_id_3";

    let root = test_env.root().join("state.json");
    let mut state_file = std::fs::File::create(&root).unwrap();
    let content = json!({
        "version": 1,
        "authentication_token": "criticalup_token_45_hahaha",
        "installations": {
            installation_id_1: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": [
                    "/path/to/proj/1/criticalup.toml",
                    "/path/to/proj/2/criticalup.toml"
                ]
            },
            installation_id_2: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": []
            },
            installation_id_3: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/rustc"
                },
                "manifests": []
            }
        }
    })
    .to_string();
    state_file.write_all(content.as_bytes()).unwrap();

    assert_output!(test_env.cmd().args(["clean"]));

    let state_file_actual: Value =
        serde_json::from_reader(BufReader::new(File::open(&root).unwrap())).unwrap();
    // "installation_id_2" is not present.
    assert_eq!(
        state_file_actual.pointer(format!("/installations/{}", installation_id_2).as_str()),
        None
    );
    // "installation_id_3" is not present.
    assert_eq!(
        state_file_actual.pointer(format!("/installations/{}", installation_id_3).as_str()),
        None
    );
    // "installation_id_1" is still present with correct values.
    assert_eq!(
        state_file_actual
            .pointer(format!("/installations/{}", installation_id_1).as_str())
            .unwrap(),
        &json!({
            "binary_proxies": {
                "cargo": "/path/toolchains/bin/cargo"
            },
            "manifests": [
                "/path/to/proj/1/criticalup.toml",
                "/path/to/proj/2/criticalup.toml"
            ]
        })
    );
}

#[test]
fn clean_deletes_only_unused_installations_also_from_disk() {
    let test_env = TestEnvironment::prepare();
    let root = test_env.root();
    let toolchains_dir = root.join("toolchains");
    fs::create_dir_all(&toolchains_dir).unwrap();

    let installation_id_1 = "installation_id_1";
    let installation_id_2 = "installation_id_2";
    let installation_id_3 = "installation_id_3";

    let state_file_in_root = root.join("state.json");
    let mut state_file = std::fs::File::create(&state_file_in_root).unwrap();
    let content = json!({
        "version": 1,
        "authentication_token": "criticalup_token_45_hahaha",
        "installations": {
            installation_id_1: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": [
                    "/path/to/proj/1/criticalup.toml",
                    "/path/to/proj/2/criticalup.toml"
                ]
            },
            installation_id_2: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": []
            },
            installation_id_3: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/rustc"
                },
                "manifests": []
            }
        }
    })
    .to_string();
    state_file.write_all(content.as_bytes()).unwrap();

    // Create the corresponding physical directories of installations.
    // TODO: We have to generate these by running `install` command, once tests for those are setup.
    fs::create_dir_all(toolchains_dir.join(installation_id_1)).unwrap();
    fs::create_dir_all(toolchains_dir.join(installation_id_2)).unwrap();
    fs::create_dir_all(toolchains_dir.join(installation_id_3)).unwrap();

    assert!(toolchains_dir.join(installation_id_1).exists());
    assert!(toolchains_dir.join(installation_id_2).exists());
    assert!(toolchains_dir.join(installation_id_3).exists());

    // Run the `clean` command.
    assert_output!(test_env.cmd().args(["clean"]));

    // Test the actual values.
    let state_file_actual: Value =
        serde_json::from_reader(BufReader::new(File::open(&state_file_in_root).unwrap())).unwrap();

    // "installation_id_2" is not present.
    assert_eq!(
        state_file_actual.pointer("/installations/installation_id_2"),
        None
    );
    // "installation_id_3" is not present.
    assert_eq!(
        state_file_actual.pointer("/installations/installation_id_3"),
        None
    );
    // "installation_id_1" is still present with correct values.
    assert_eq!(
        state_file_actual
            .pointer("/installations/installation_id_1")
            .unwrap(),
        &json!({
            "binary_proxies": {
                "cargo": "/path/toolchains/bin/cargo"
            },
            "manifests": [
                "/path/to/proj/1/criticalup.toml",
                "/path/to/proj/2/criticalup.toml"
            ]
        })
    );

    assert!(toolchains_dir.join(installation_id_1).exists());
    assert!(!toolchains_dir.join(installation_id_2).exists()); // Does not exist.
    assert!(!toolchains_dir.join(installation_id_3).exists()); // Does not exist.
}

#[test]
fn removes_unused_installations_from_disk_that_do_not_have_state() {
    let test_env = TestEnvironment::prepare();
    let root = test_env.root();
    let toolchains_dir = root.join("toolchains");
    fs::create_dir_all(&toolchains_dir).unwrap();

    let installation_id_1 = "installation_id_1";
    let installation_id_2 = "installation_id_2";
    let installation_id_3 = "installation_id_3"; // No State, only directory

    let state_file_in_root = root.join("state.json");
    let mut state_file = std::fs::File::create(&state_file_in_root).unwrap();
    let content = json!({
        "version": 1,
        "authentication_token": "criticalup_token_45_hahaha",
        "installations": {
            installation_id_1: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": [
                    "/path/to/proj/1/criticalup.toml",
                    "/path/to/proj/2/criticalup.toml"
                ]
            },
            installation_id_2: {
                "binary_proxies": {
                    "cargo": "/path/toolchains/bin/cargo"
                },
                "manifests": []
            }
        }
    })
    .to_string();
    state_file.write_all(content.as_bytes()).unwrap();

    // Create the corresponding physical directories of installations.
    // TODO: We have to generate these by running `install` command, once tests for those are setup.
    fs::create_dir_all(toolchains_dir.join(installation_id_1)).unwrap();
    fs::create_dir_all(toolchains_dir.join(installation_id_2)).unwrap();
    fs::create_dir_all(toolchains_dir.join(installation_id_3)).unwrap();

    assert!(toolchains_dir.join(installation_id_1).exists());
    assert!(toolchains_dir.join(installation_id_2).exists());
    assert!(toolchains_dir.join(installation_id_3).exists());

    // Run the `clean` command.
    assert_output!(test_env.cmd().args(["clean"]));

    // Test the actual values.
    let state_file_actual: Value =
        serde_json::from_reader(BufReader::new(File::open(&state_file_in_root).unwrap())).unwrap();

    // "installation_id_2" is not present.
    assert_eq!(
        state_file_actual.pointer("/installations/installation_id_2"),
        None
    );
    // "installation_id_1" is still present with correct values.
    assert_eq!(
        state_file_actual
            .pointer("/installations/installation_id_1")
            .unwrap(),
        &json!({
            "binary_proxies": {
                "cargo": "/path/toolchains/bin/cargo"
            },
            "manifests": [
                "/path/to/proj/1/criticalup.toml",
                "/path/to/proj/2/criticalup.toml"
            ]
        })
    );

    assert!(toolchains_dir.join(installation_id_1).exists());
    assert!(!toolchains_dir.join(installation_id_2).exists()); // Does not exist.
    assert!(!toolchains_dir.join(installation_id_3).exists()); // Does not exist.
}
