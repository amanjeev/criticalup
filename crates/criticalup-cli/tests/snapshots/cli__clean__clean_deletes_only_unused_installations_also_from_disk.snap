---
source: crates/criticalup-cli/tests/cli/clean.rs
expression: repr
---
exit: exit status: 0

empty stdout

stderr
------
 INFO Deleting unused installation installation_id_2
 INFO Deleting unused installation directory /path/to/toolchain/installation/installation_id_2/
 INFO Deleting unused installation installation_id_3
 INFO Deleting unused installation directory /path/to/toolchain/installation/installation_id_3/
 INFO no untracked installation directories found
------
